--- a/net/minecraft/server/ReloadableServerRegistries.java
+++ b/net/minecraft/server/ReloadableServerRegistries.java
@@ -43,8 +_,9 @@
         List<HolderLookup.RegistryLookup<?>> list = TagLoader.buildUpdatedLookups(p_335950_.getAccessForLoading(RegistryLayer.RELOADABLE), p_363295_);
         HolderLookup.Provider holderlookup$provider = HolderLookup.Provider.create(list.stream());
         RegistryOps<JsonElement> registryops = holderlookup$provider.createSerializationContext(JsonOps.INSTANCE);
+        final io.papermc.paper.registry.data.util.Conversions conversions = new io.papermc.paper.registry.data.util.Conversions(registryops.lookupProvider); // Paper
         List<CompletableFuture<WritableRegistry<?>>> list1 = LootDataType.values()
-            .map(p_359506_ -> scheduleRegistryLoad((LootDataType<?>)p_359506_, registryops, p_335786_, p_335516_))
+            .map(p_359506_ -> scheduleRegistryLoad((LootDataType<?>)p_359506_, registryops, p_335786_, p_335516_, conversions)) // Paper
             .toList();
         CompletableFuture<List<WritableRegistry<?>>> completablefuture = Util.sequence(list1);
         return completablefuture.thenApplyAsync(
@@ -53,19 +_,29 @@
     }
 
     private static <T> CompletableFuture<WritableRegistry<?>> scheduleRegistryLoad(
-        LootDataType<T> p_335741_, RegistryOps<JsonElement> p_336173_, ResourceManager p_335893_, Executor p_336104_
+        LootDataType<T> p_335741_, RegistryOps<JsonElement> p_336173_, ResourceManager p_335893_, Executor p_336104_, io.papermc.paper.registry.data.util.Conversions conversions // Paper
     ) {
         return CompletableFuture.supplyAsync(
             () -> {
                 WritableRegistry<T> writableregistry = new MappedRegistry<>(p_335741_.registryKey(), Lifecycle.experimental());
+                io.papermc.paper.registry.PaperRegistryAccess.instance().registerReloadableRegistry(writableregistry); // Paper - register reloadable registry
                 Map<ResourceLocation, T> map = new HashMap<>();
-                SimpleJsonResourceReloadListener.scanDirectory(p_335893_, p_335741_.registryKey(), p_336173_, p_335741_.codec(), map);
+                var provider = net.neoforged.neoforge.common.CommonHooks.extractLookupProvider(p_336173_);
+                Map<ResourceLocation, Optional<T>> optionalMap = new HashMap<>();
+                SimpleJsonResourceReloadListener.scanDirectoryWithOptionalValues(p_335893_, p_335741_.registryKey(), p_336173_, p_335741_.conditionalCodec(), optionalMap);
+                optionalMap.forEach((rl, optionalEntry) -> {
+                    optionalEntry.ifPresent(entry -> p_335741_.idSetter().accept(entry, rl));
+                    T value = optionalEntry.orElse(p_335741_.defaultValue());
+                    if (value instanceof LootTable lootTable) value = (T) net.neoforged.neoforge.event.EventHooks.loadLootTable(provider, rl, lootTable);
+                    if (value != null)
+                        map.put(rl, value);
+                });
                 map.forEach(
-                    (p_335721_, p_335683_) -> writableregistry.register(
-                        ResourceKey.create(p_335741_.registryKey(), p_335721_), (T)p_335683_, DEFAULT_REGISTRATION_INFO
+                    (p_335721_, p_335683_) -> io.papermc.paper.registry.PaperRegistryListenerManager.INSTANCE.registerWithListeners(writableregistry, // Paper - register with listeners
+                        ResourceKey.create(p_335741_.registryKey(), p_335721_), (T)p_335683_, DEFAULT_REGISTRATION_INFO, conversions // Paper - register with listeners
                     )
                 );
-                TagLoader.loadTagsForRegistry(p_335893_, writableregistry);
+                TagLoader.loadTagsForRegistry(p_335893_, writableregistry, io.papermc.paper.plugin.lifecycle.event.registrar.ReloadableRegistrarEvent.Cause.RELOAD); // Paper - tag life cycle - reload
                 return writableregistry;
             },
             p_336104_
