--- a/net/minecraft/world/level/CommandBlockListenerAbstract.java
+++ b/net/minecraft/world/level/CommandBlockListenerAbstract.java
@@ -32,6 +32,9 @@
     IChatBaseComponent lastOutput;
     private String command = "";
     private @Nullable IChatBaseComponent customName;
+    // CraftBukkit start
+    public abstract org.bukkit.command.CommandSender getBukkitSender(CommandListenerWrapper wrapper);
+    // CraftBukkit end
 
     public CommandBlockListenerAbstract() {}
 
@@ -114,7 +117,7 @@
 
                         });
 
-                        worldserver.getServer().getCommands().performPrefixedCommand(commandlistenerwrapper, this.command);
+                        worldserver.getServer().getCommands().dispatchServerCommand(commandlistenerwrapper, this.command); // CraftBukkit
                     }
                 } catch (Throwable throwable) {
                     CrashReport crashreport = CrashReport.forThrowable(throwable, "Executing command block");
@@ -139,7 +142,11 @@
     }
 
     public CommandBlockListenerAbstract.@Nullable CloseableCommandBlockSource createSource(WorldServer worldserver) {
-        return this.trackOutput ? new CommandBlockListenerAbstract.CloseableCommandBlockSource(worldserver) : null;
+        // CraftBukkit start - don't use a null source with no output as we still need the Bukkit sender, just close it immediately to give the same effect
+        CommandBlockListenerAbstract.CloseableCommandBlockSource source = new CommandBlockListenerAbstract.CloseableCommandBlockSource(worldserver);
+        source.closed = !this.trackOutput;
+        return source;
+        // CraftBukkit end
     }
 
     public IChatBaseComponent getName() {
@@ -182,6 +189,13 @@
             this.level = worldserver;
         }
 
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandListenerWrapper wrapper) {
+            return CommandBlockListenerAbstract.this.getBukkitSender(wrapper);
+        }
+        // CraftBukkit end
+
         @Override
         public boolean acceptsSuccess() {
             return !this.closed && (Boolean) this.level.getGameRules().get(GameRules.SEND_COMMAND_FEEDBACK);
