--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -96,13 +145,61 @@
     @Override
     public void handlePong(ServerboundPongPacket serverboundPongPacket) {}
 
+    // CraftBukkit start
+    private static final Identifier CUSTOM_REGISTER = Identifier.withDefaultNamespace("register");
+    private static final Identifier CUSTOM_UNREGISTER = Identifier.withDefaultNamespace("unregister");
+
     @Override
-    public void handleCustomPayload(ServerboundCustomPayloadPacket packet) {}
+    public void handleCustomPayload(ServerboundCustomPayloadPacket packet) {
+        if (!(packet.payload() instanceof DiscardedPayload)) {
+            return;
+        }
+        PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
+        Identifier identifier = packet.payload().type().id();
+        ByteBuf payload = ((DiscardedPayload)packet.payload()).data();
+
+        if (identifier.equals(CUSTOM_REGISTER)) {
+            try {
+                String channels = payload.toString(com.google.common.base.Charsets.UTF_8);
+                for (String channel : channels.split("\0")) {
+                    getCraftPlayer().addChannel(channel);
+                }
+            } catch (Exception ex) {
+                ServerCommonPacketListenerImpl.LOGGER.error("Couldn\'t register custom payload", ex);
+                this.disconnect(Component.literal("Invalid payload REGISTER!"));
+            }
+        } else if (identifier.equals(CUSTOM_UNREGISTER)) {
+            try {
+                String channels = payload.toString(com.google.common.base.Charsets.UTF_8);
+                for (String channel : channels.split("\0")) {
+                    getCraftPlayer().removeChannel(channel);
+                }
+            } catch (Exception ex) {
+                ServerCommonPacketListenerImpl.LOGGER.error("Couldn\'t unregister custom payload", ex);
+                this.disconnect(Component.literal("Invalid payload UNREGISTER!"));
+            }
+        } else {
+            try {
+                byte[] data = new byte[payload.readableBytes()];
+                payload.readBytes(data);
+                cserver.getMessenger().dispatchIncomingMessage(player.getBukkitEntity(), identifier.toString(), data);
+            } catch (Exception ex) {
+                ServerCommonPacketListenerImpl.LOGGER.error("Couldn\'t dispatch custom payload", ex);
+                this.disconnect(Component.literal("Invalid custom payload!"));
+            }
+        }
+
+    }
