--- a/net/minecraft/world/level/block/BlockIce.java
+++ b/net/minecraft/world/level/block/BlockIce.java
@@ -61,6 +61,11 @@
     }
 
     protected void melt(IBlockData iblockdata, World world, BlockPosition blockposition) {
+        // CraftBukkit start
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, blockposition, world.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, blockposition) ? Blocks.AIR.defaultBlockState() : meltsInto()).isCancelled()) {
+            return;
+        }
+        // CraftBukkit end
         if ((Boolean) world.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, blockposition)) {
             world.removeBlock(blockposition, false);
         } else {
