--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -34,7 +_,7 @@
     FluidState getFluidState(BlockPos p_45569_);
 
     default int getLightEmission(BlockPos p_45572_) {
-        return this.getBlockState(p_45572_).getLightEmission();
+        return this.getBlockState(p_45572_).getLightEmission(this, p_45572_);
     }
 
     default Stream<BlockState> getBlockStates(AABB p_45557_) {
@@ -61,6 +_,23 @@
             }
         );
     }
+
+    // CraftBukkit start - moved block handling into separate method for use by Block#rayTrace
+    default BlockHitResult clip(ClipContext raytrace1, BlockPos blockposition) {
+        BlockState iblockdata = this.getBlockState(blockposition);
+        FluidState fluid = this.getFluidState(blockposition);
+        Vec3 vec3d = raytrace1.getFrom();
+        Vec3 vec3d1 = raytrace1.getTo();
+        VoxelShape voxelshape = raytrace1.getBlockShape(iblockdata, this, blockposition);
+        BlockHitResult movingobjectpositionblock = this.clipWithInteractionOverride(vec3d, vec3d1, blockposition, voxelshape, iblockdata);
+        VoxelShape voxelshape1 = raytrace1.getFluidShape(fluid, this, blockposition);
+        BlockHitResult movingobjectpositionblock1 = voxelshape1.clip(vec3d, vec3d1, blockposition);
+        double d0 = movingobjectpositionblock == null ? Double.MAX_VALUE : raytrace1.getFrom().distanceToSqr(movingobjectpositionblock.getLocation());
+        double d1 = movingobjectpositionblock1 == null ? Double.MAX_VALUE : raytrace1.getFrom().distanceToSqr(movingobjectpositionblock1.getLocation());
+
+        return d0 <= d1 ? movingobjectpositionblock : movingobjectpositionblock1;
+    }
+    // CraftBukkit end
 
     default BlockHitResult clip(ClipContext p_45548_) {
         return traverseBlocks(p_45548_.getFrom(), p_45548_.getTo(), p_45548_, (p_151359_, p_151360_) -> {
