--- a/net/minecraft/world/level/TicketStorage.java
+++ b/net/minecraft/world/level/TicketStorage.java
@@ -36,7 +_,8 @@
     private static final Codec<Pair<ChunkPos, Ticket>> TICKET_ENTRY = Codec.mapPair(ChunkPos.CODEC.fieldOf("chunk_pos"), Ticket.CODEC).codec();
     public static final Codec<TicketStorage> CODEC = RecordCodecBuilder.create(
         p_400948_ -> p_400948_.group(TICKET_ENTRY.listOf().optionalFieldOf("tickets", List.of()).forGetter(TicketStorage::packTickets))
-            .apply(p_400948_, TicketStorage::fromPacked)
+            .and(net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.defineExtraStorageParams())
+            .apply(p_400948_, (tickets, neoData) -> net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.readStoredTickets(TicketStorage::fromPacked, tickets, neoData))
     );
     public static final SavedDataType<TicketStorage> TYPE = new SavedDataType<>("chunks", TicketStorage::new, CODEC, DataFixTypes.SAVED_DATA_FORCED_CHUNKS);
     public final Long2ObjectOpenHashMap<List<Ticket>> tickets;
@@ -51,6 +_,7 @@
         this.tickets = p_393873_;
         this.deactivatedTickets = p_394615_;
         this.updateForcedChunks();
+        this.updateForcedNaturalSpawning();
     }
 
     public TicketStorage() {
@@ -174,13 +_,14 @@
         if (p_394469_.getType().equals(TicketType.FORCED)) {
             this.chunksWithForcedTickets.add(p_394247_);
         }
+        if (p_394469_.getType().forceNaturalSpawning()) chunksWithForceNaturalSpawning.add(p_394247_);
 
         this.setDirty();
         return true;
     }
 
-    private static boolean isTicketSameTypeAndLevel(Ticket p_394344_, Ticket p_394181_) {
-        return p_394181_.getType() == p_394344_.getType() && p_394181_.getTicketLevel() == p_394344_.getTicketLevel();
+    private static boolean isTicketSameTypeAndLevel(Ticket first, Ticket second) {
+        return second.getType() == first.getType() && second.getTicketLevel() == first.getTicketLevel() && java.util.Objects.equals(second.getIdentifier(), first.getIdentifier()); // Paper - add identifier
     }
 
     public int getTicketLevelAt(long p_393578_, boolean p_393891_) {
@@ -261,6 +_,7 @@
                 if (p_394054_.getType().equals(TicketType.FORCED)) {
                     this.updateForcedChunks();
                 }
+                if (p_394054_.getType().forceNaturalSpawning()) this.updateForcedNaturalSpawning();
 
                 this.setDirty();
                 return true;
@@ -269,7 +_,7 @@
     }
 
     private void updateForcedChunks() {
-        this.chunksWithForcedTickets = this.getAllChunksWithTicketThat(p_393889_ -> p_393889_.getType().equals(TicketType.FORCED));
+        throw new UnsupportedOperationException(); // Paper - rewrite chunk system
     }
 
     public String getTicketDebugString(long p_393749_, boolean p_394364_) {
@@ -302,12 +_,13 @@
     }
 
     public void deactivateTicketsOnClosing() {
-        this.removeTicketIf((p_394604_, p_435717_) -> p_394604_.getType() != TicketType.UNKNOWN, this.deactivatedTickets);
+        // Paper - rewrite chunk system
     }
 
     public void removeTicketIf(TicketStorage.TicketPredicate p_435606_, @Nullable Long2ObjectOpenHashMap<List<Ticket>> p_393746_) {
         ObjectIterator<Entry<List<Ticket>>> objectiterator = this.tickets.long2ObjectEntrySet().fastIterator();
         boolean flag = false;
+        boolean updateNaturalSpawning = false;
 
         while (objectiterator.hasNext()) {
             Entry<List<Ticket>> entry = objectiterator.next();
@@ -336,6 +_,7 @@
                     if (ticket.getType().equals(TicketType.FORCED)) {
                         flag = true;
                     }
+                    if (ticket.getType().forceNaturalSpawning()) updateNaturalSpawning = true;
                 }
             }
 
@@ -358,6 +_,7 @@
         if (flag) {
             this.updateForcedChunks();
         }
+        if (updateNaturalSpawning) this.updateForcedNaturalSpawning();
     }
 
     public void replaceTicketLevelOfType(int p_393875_, TicketType p_394261_) {
@@ -411,5 +_,42 @@
 
     public interface TicketPredicate {
         boolean test(Ticket p_434782_, long p_434387_);
+    }
+
+    // Paper start
+    public boolean addPluginRegionTicket(final ChunkPos pos, final org.bukkit.plugin.Plugin value) {
+        // Keep inline with force loading
+        return addTicket(pos.toLong(), new Ticket(TicketType.PLUGIN_TICKET, ChunkMap.FORCED_TICKET_LEVEL, value));
+    }
+
+    public boolean removePluginRegionTicket(final ChunkPos pos, final org.bukkit.plugin.Plugin value) {
+        // Keep inline with force loading
+        return removeTicket(pos.toLong(), new Ticket(TicketType.PLUGIN_TICKET, ChunkMap.FORCED_TICKET_LEVEL, value));
+    }
+
+    public void removeAllPluginRegionTickets(TicketType ticketType, int ticketLevel, org.bukkit.plugin.Plugin ticketIdentifier) {
+        this.chunkMap.level.moonrise$getChunkTaskScheduler().chunkHolderManager.removeAllTicketsFor(ticketType, ticketLevel, ticketIdentifier); // Paper - rewrite chunk system
+    }
+    // Paper end
+
+    // Neo: Keep track of forced loaded chunks caused by entities or blocks.
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> blockForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>(this, net.neoforged.neoforge.common.NeoForgeMod.BLOCK_TICKET, net.neoforged.neoforge.common.NeoForgeMod.BLOCK_WITH_NATURAL_SPAWNING_TICKET);
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> entityForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>(this, net.neoforged.neoforge.common.NeoForgeMod.ENTITY_TICKET, net.neoforged.neoforge.common.NeoForgeMod.ENTITY_WITH_NATURAL_SPAWNING_TICKET);
+    private LongSet chunksWithForceNaturalSpawning = new LongOpenHashSet();
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> getBlockForcedChunks() {
+        return this.blockForcedChunks;
+    }
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> getEntityForcedChunks() {
+        return this.entityForcedChunks;
+    }
+
+    private void updateForcedNaturalSpawning() {
+        this.chunksWithForceNaturalSpawning = this.getAllChunksWithTicketThat(ticket -> ticket.getType().forceNaturalSpawning());
+    }
+
+    public boolean shouldForceNaturalSpawning(ChunkPos chunkPos) {
+        return chunksWithForceNaturalSpawning.contains(chunkPos.toLong());
     }
 }
