--- a/net/minecraft/world/entity/projectile/arrow/EntityThrownTrident.java
+++ b/net/minecraft/world/entity/projectile/arrow/EntityThrownTrident.java
@@ -28,6 +28,10 @@
 import net.minecraft.world.phys.Vec3D;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityThrownTrident extends EntityArrow {
 
     private static final DataWatcherObject<Byte> ID_LOYALTY = DataWatcher.<Byte>defineId(EntityThrownTrident.class, DataWatcherRegistry.BYTE);
@@ -81,10 +85,10 @@
                     }
                 }
 
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
             } else {
                 if (!(entity instanceof EntityHuman) && this.position().distanceTo(entity.getEyePosition()) < (double) entity.getBbWidth() + 1.0D) {
-                    this.discard();
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                     return;
                 }
 
