--- a/net/minecraft/world/level/levelgen/structure/PersistentStructureLegacy.java
+++ b/net/minecraft/world/level/levelgen/structure/PersistentStructureLegacy.java
@@ -27,6 +27,10 @@
 import net.minecraft.world.level.storage.WorldPersistentData;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.level.dimension.WorldDimension;
+// CraftBukkit end
+
 public class PersistentStructureLegacy implements LegacyTagFixer {
 
     public static final int LAST_MONOLYTH_STRUCTURE_DATA_VERSION = 1493;
@@ -216,7 +220,7 @@
                             Optional<String> optional = nbttaglist.getCompound(0).flatMap((nbttagcompound2) -> {
                                 return nbttagcompound2.getString("id");
                             });
-                            Map map = PersistentStructureLegacy.LEGACY_TO_CURRENT_MAP;
+                            Map<String, String> map = PersistentStructureLegacy.LEGACY_TO_CURRENT_MAP; // CraftBukkit - decompile error
 
                             Objects.requireNonNull(map);
                             optional.map(map::get).ifPresent((s2) -> {
@@ -253,18 +257,18 @@
         }
     }
 
-    public static Supplier<LegacyTagFixer> getLegacyTagFixer(ResourceKey<World> resourcekey, Supplier<@Nullable WorldPersistentData> supplier, DataFixer datafixer) {
-        if (resourcekey == World.OVERWORLD) {
+    public static Supplier<LegacyTagFixer> getLegacyTagFixer(ResourceKey<WorldDimension> resourcekey, Supplier<@Nullable WorldPersistentData> supplier, DataFixer datafixer) { // CraftBukkit
+        if (resourcekey == WorldDimension.OVERWORLD) { // CraftBukkit
             return () -> {
                 return new PersistentStructureLegacy((WorldPersistentData) supplier.get(), ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"), datafixer);
             };
-        } else if (resourcekey == World.NETHER) {
+        } else if (resourcekey == WorldDimension.NETHER) { // CraftBukkit
             List<String> list = ImmutableList.of("Fortress");
 
             return () -> {
                 return new PersistentStructureLegacy((WorldPersistentData) supplier.get(), list, list, datafixer);
             };
-        } else if (resourcekey == World.END) {
+        } else if (resourcekey == WorldDimension.END) { // CraftBukkit
             List<String> list1 = ImmutableList.of("EndCity");
 
             return () -> {
