--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -77,7 +_,7 @@
     private static final Identifier ZOMBIE_RANDOM_SPAWN_BONUS_ID = Identifier.withDefaultNamespace("zombie_random_spawn_bonus");
     private static final EntityDataAccessor<Boolean> DATA_BABY_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Integer> DATA_SPECIAL_TYPE_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.INT);
-    private static final EntityDataAccessor<Boolean> DATA_DROWNED_CONVERSION_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.BOOLEAN);
+    public static final EntityDataAccessor<Boolean> DATA_DROWNED_CONVERSION_ID = SynchedEntityData.defineId(Zombie.class, EntityDataSerializers.BOOLEAN);
     public static final float ZOMBIE_LEADER_CHANCE = 0.05F;
     public static final int REINFORCEMENT_ATTEMPTS = 50;
     public static final int REINFORCEMENT_RANGE_MAX = 40;
@@ -92,7 +_,7 @@
     private final BreakDoorGoal breakDoorGoal = new BreakDoorGoal(this, DOOR_BREAKING_PREDICATE);
     private boolean canBreakDoors = false;
     private int inWaterTime = 0;
-    protected int conversionTime;
+    public int conversionTime;
 
     public Zombie(EntityType<? extends Zombie> p_481365_, Level p_479230_) {
         super(p_481365_, p_479230_);
@@ -226,12 +_,13 @@
         super.tick();
     }
 
-    private void startUnderWaterConversion(int p_479096_) {
+    public void startUnderWaterConversion(int p_479096_) {
         this.conversionTime = p_479096_;
         this.getEntityData().set(DATA_DROWNED_CONVERSION_ID, true);
     }
 
     protected void doUnderWaterConversion(ServerLevel p_481555_) {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.DROWNED, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(p_481555_, EntityType.DROWNED);
         if (!this.isSilent()) {
             p_481555_.levelEvent(null, 1040, this.blockPosition(), 0);
@@ -242,7 +_,10 @@
         this.convertTo(
             p_482190_,
             ConversionParams.single(this, true, true),
-            p_480962_ -> p_480962_.handleAttributes(p_481352_.getCurrentDifficultyAt(p_480962_.blockPosition()).getSpecialMultiplier())
+            p_381517_ -> {
+                p_381517_.handleAttributes(p_481352_.getCurrentDifficultyAt(p_381517_.blockPosition()).getSpecialMultiplier());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_381517_);
+            }
         );
     }
 
@@ -262,6 +_,7 @@
                 p_480487_.setGossips(p_479207_.getGossips().copy());
                 p_480487_.setTradeOffers(p_479207_.getOffers().copy());
                 p_480487_.setVillagerXp(p_479207_.getVillagerXp());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(p_479207_, p_480487_);
                 if (!this.isSilent()) {
                     p_480405_.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -417,7 +_,7 @@
     @Override
     public boolean killedEntity(ServerLevel p_479575_, LivingEntity p_478394_, DamageSource p_481123_) {
         boolean flag = super.killedEntity(p_479575_, p_478394_, p_481123_);
-        if ((p_479575_.getDifficulty() == Difficulty.NORMAL || p_479575_.getDifficulty() == Difficulty.HARD) && p_478394_ instanceof Villager villager) {
+        if ((p_479575_.getDifficulty() == Difficulty.NORMAL || p_479575_.getDifficulty() == Difficulty.HARD) && p_478394_ instanceof Villager villager && net.neoforged.neoforge.event.EventHooks.canLivingConvert(p_478394_, EntityType.ZOMBIE_VILLAGER, (timer) -> {})) {
             if (p_479575_.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
                 return flag;
             }
