--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -4,6 +_,7 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.ImmutableMap.Builder;
 import com.google.common.net.HostAndPort;
+import com.mohistmc.youer.YouerConfig;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import io.netty.handler.ssl.SslContext;
@@ -73,6 +_,16 @@
 import net.minecraft.world.level.gamerules.GameRules;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.io.IoBuilder;
+import org.bukkit.craftbukkit.util.TerminalConsoleWriterThread;
+import org.bukkit.event.server.RemoteServerCommandEvent;
+import org.bukkit.event.server.ServerCommandEvent;
+import org.jline.reader.EndOfFileException;
+import org.jline.reader.LineReader;
+import org.jline.reader.LineReaderBuilder;
+import org.jline.reader.UserInterruptException;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -82,17 +_,18 @@
     private static final int CONVERSION_RETRIES = 2;
     private final List<ConsoleInput> consoleInput = Collections.synchronizedList(Lists.newArrayList());
     private @Nullable QueryThreadGs4 queryThreadGs4;
-    private final RconConsoleSource rconConsoleSource;
+    // private final RconConsoleSource rconConsoleSource;
     private @Nullable RconThread rconThread;
-    private final DedicatedServerSettings settings;
+    public final DedicatedServerSettings settings;
     private @Nullable MinecraftServerGui gui;
     private final @Nullable ServerTextFilter serverTextFilter;
     private @Nullable RemoteSampleLogger tickTimeLogger;
     private boolean isTickTimeLoggingEnabled;
-    private final ServerLinks serverLinks;
+    public ServerLinks serverLinks;
     private final Map<String, String> codeOfConductTexts;
     private @Nullable ManagementServer jsonRpcServer;
     private long lastHeartbeat;
+    private net.neoforged.neoforge.server.dedicated.@Nullable DedicatedLanServerPinger dediLanPinger;
 
     public DedicatedServer(
         Thread p_214789_,
@@ -105,7 +_,7 @@
     ) {
         super(p_214789_, p_214790_, p_214791_, p_214792_, Proxy.NO_PROXY, p_214794_, p_214795_, LoggingLevelLoadListener.forDedicatedServer());
         this.settings = p_214793_;
-        this.rconConsoleSource = new RconConsoleSource(this);
+        // this.rconConsoleSource = new RconConsoleSource(this); // CraftBukkit - remove field
         this.serverTextFilter = ServerTextFilter.createFromConfig(p_214793_.getProperties());
         this.serverLinks = createServerLinks(p_214793_);
         if (p_214793_.getProperties().codeOfConduct) {
@@ -188,21 +_,81 @@
             }
         }
 
+        // CraftBukkit start
+        LineReader reader = LineReaderBuilder.builder()
+                .terminal(terminal)
+                .option(LineReader.Option.DISABLE_EVENT_EXPANSION, true)
+                .completer(new org.bukkit.craftbukkit.command.ConsoleCommandCompleter())
+                .build();
+        // CraftBukkit end
         Thread thread1 = new Thread("Server console handler") {
             @Override
             public void run() {
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.Main.useConsole) {
+                    return;
+                }
+
+                // MC-33041, SPIGOT-5538: if System.in is not valid due to javaw, then return
+                try {
+                    System.in.available();
+                } catch (IOException ex) {
+                    return;
+                }
+                // CraftBukkit end
+                if (net.neoforged.neoforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
                 BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
 
                 String s4;
                 try {
                     while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning() && (s4 = bufferedreader.readLine()) != null) {
-                        DedicatedServer.this.handleConsoleInput(s4, DedicatedServer.this.createCommandSourceStack());
+                        try {
+                            if (org.bukkit.craftbukkit.Main.useJline) {
+                                s4 = reader.readLine(">", null);
+                            } else {
+                                s4 = reader.readLine();
+                            }
+
+                            if (s4.trim().length() > 0) { // Trim to filter lines which are just spaces
+                                DedicatedServer.this.handleConsoleInput(s4, DedicatedServer.this.createCommandSourceStack());
+                            }
+                        } catch (EndOfFileException | UserInterruptException eof) {
+                            // SPIGOT-5220: Throttle if EOF (ctrl^d) or stdin is /dev/null
+                            try {
+                                Thread.sleep(50L);
+                            } catch (InterruptedException ex) {
+                                Thread.currentThread().interrupt();
+                            }
+                        }
                     }
                 } catch (IOException ioexception1) {
                     DedicatedServer.LOGGER.error("Exception handling console input", (Throwable)ioexception1);
                 }
             }
         };
+
+        // CraftBukkit start - TODO: handle command-line logging arguments
+        java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
+        global.setUseParentHandlers(false);
+        for (java.util.logging.Handler handler : global.getHandlers()) {
+            global.removeHandler(handler);
+        }
+        global.addHandler(new org.bukkit.craftbukkit.util.ForwardLogHandler());
+
+        final org.apache.logging.log4j.core.Logger logger = ((org.apache.logging.log4j.core.Logger) LogManager.getRootLogger());
+        for (org.apache.logging.log4j.core.Appender appender : logger.getAppenders().values()) {
+            if (appender instanceof org.apache.logging.log4j.core.appender.ConsoleAppender) {
+                logger.removeAppender(appender);
+            }
+        }
+
+        TerminalConsoleWriterThread writerThread = new TerminalConsoleWriterThread(System.out, reader);
+        writerThread.start();
+
+        System.setOut(IoBuilder.forLogger(logger).setLevel(Level.INFO).buildPrintStream());
+        System.setErr(IoBuilder.forLogger(logger).setLevel(Level.WARN).buildPrintStream());
+        // CraftBukkit end
+
         thread1.setDaemon(true);
         thread1.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
         thread1.start();
@@ -220,7 +_,15 @@
             this.setPreventProxyConnections(dedicatedserverproperties.preventProxyConnections);
             this.setLocalIp(dedicatedserverproperties.serverIp);
         }
-
+        this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+        // Spigot start
+        org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
+        org.spigotmc.SpigotConfig.registerCommands();
+        // Spigot end
+        // Youer start
+        YouerConfig.init((java.io.File) options.valueOf("youer-settings"));
+        YouerConfig.registerCommands();
+        // Youer end
         this.worldData.setGameType(dedicatedserverproperties.gameMode.get());
         LOGGER.info("Default game type: {}", dedicatedserverproperties.gameMode.get());
         InetAddress inetaddress = null;
@@ -244,6 +_,11 @@
             return false;
         }
 
+        // CraftBukkit start
+        server.loadPlugins();
+        server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+        // CraftBukkit end
+
         if (!this.usesAuthentication()) {
             LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
             LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
@@ -260,15 +_,16 @@
         if (!OldUsersConverter.serverReadyAfterUserconversion(this)) {
             return false;
         } else {
-            this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
             this.tickTimeLogger = new RemoteSampleLogger(TpsDebugDimensions.values().length, this.debugSubscribers(), RemoteDebugSampleType.TICK_TIME);
             long j = Util.getNanos();
             this.services.nameToIdCache().resolveOfflineUsers(!this.usesAuthentication());
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
             LOGGER.info("Preparing level \"{}\"", this.getLevelIdName());
             this.loadLevel();
             long k = Util.getNanos() - j;
             String s3 = String.format(Locale.ROOT, "%.3fs", k / 1.0E9);
             LOGGER.info("Done ({})! For help, type \"help\"", s3);
+            this.nextTickTimeNanos = Util.getNanos(); // Neo: Update server time to prevent watchdog/spamming during long load.
             if (dedicatedserverproperties.announcePlayerAchievements != null) {
                 this.worldData.getGameRules().set(GameRules.SHOW_ADVANCEMENT_MESSAGES, dedicatedserverproperties.announcePlayerAchievements, this);
             }
@@ -297,6 +_,12 @@
             }
 
             this.notificationManager().serverStarted();
+            if (net.neoforged.neoforge.common.config.NeoForgeServerConfig.INSTANCE.advertiseDedicatedServerToLan.get()) {
+                this.dediLanPinger = new net.neoforged.neoforge.server.dedicated.DedicatedLanServerPinger(this.getMotd(), String.valueOf(this.getServerPort()));
+                this.dediLanPinger.start();
+            }
+
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
             return true;
         }
     }
@@ -431,6 +_,11 @@
             this.queryThreadGs4.stop();
         }
 
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
+
         if (this.jsonRpcServer != null) {
             try {
                 this.jsonRpcServer.stop(true);
@@ -453,7 +_,15 @@
     public void handleConsoleInputs() {
         while (!this.consoleInput.isEmpty()) {
             ConsoleInput consoleinput = this.consoleInput.remove(0);
-            this.getCommands().performPrefixedCommand(consoleinput.source, consoleinput.msg);
+            // CraftBukkit start - ServerCommand for preprocessing
+            ServerCommandEvent event = new ServerCommandEvent(console, consoleinput.msg);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) continue;
+            consoleinput = new ConsoleInput(event.getCommand(), consoleinput.source);
+
+            // this.getCommands().performPrefixedCommand(servercommand.source, servercommand.msg); // Called in dispatchServerCommand
+            server.dispatchServerCommand(console, consoleinput);
+            // CraftBukkit end
         }
     }
 
@@ -712,15 +_,35 @@
 
     @Override
     public String runCommand(String p_139644_) {
-        this.rconConsoleSource.prepareForCommand();
-        this.executeBlocking(() -> this.getCommands().performPrefixedCommand(this.rconConsoleSource.createCommandSourceStack(), p_139644_));
-        return this.rconConsoleSource.getCommandResponse();
+        // CraftBukkit start - fire RemoteServerCommandEvent
+        throw new UnsupportedOperationException("Not supported - remote source required.");
+    }
+
+    public String runCommand(RconConsoleSource rconConsoleSource, String s) {
+        rconConsoleSource.prepareForCommand();
+        this.executeBlocking(() -> {
+            CommandSourceStack wrapper = rconConsoleSource.createCommandSourceStack();
+            RemoteServerCommandEvent event = new RemoteServerCommandEvent(rconConsoleSource.getBukkitSender(wrapper), s);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            ConsoleInput serverCommand = new ConsoleInput(event.getCommand(), wrapper);
+            server.dispatchServerCommand(event.getSender(), serverCommand);
+        });
+        return rconConsoleSource.getCommandResponse();
+        // CraftBukkit end
     }
 
     @Override
     public void stopServer() {
         this.notificationManager().serverShuttingDown();
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         super.stopServer();
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
         Util.shutdownExecutors();
     }
 
