--- a/net/minecraft/nbt/ListTag.java
+++ b/net/minecraft/nbt/ListTag.java
@@ -153,7 +_,16 @@
         this(new ArrayList<>());
     }
 
-    ListTag(List<Tag> p_128721_) {
+    /**
+     * Neo: create a list tag with the given initial capacity.
+     *
+     * @param initialCapacity the initial capacity of the list tag
+     */
+    public ListTag(int initialCapacity) {
+        this(new ArrayList<>(initialCapacity));
+    }
+
+    public ListTag(List<Tag> p_128721_) {
         this.list = p_128721_;
     }
 
@@ -196,7 +_,7 @@
     }
 
     @VisibleForTesting
-    byte identifyRawElementType() {
+    public byte identifyRawElementType() {
         byte b0 = 0;
 
         for (Tag tag : this.list) {
