--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -43,6 +43,19 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.core.Registry;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.dimension.LevelStem;
+import net.minecraft.world.level.levelgen.WorldDimensions;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
+// CraftBukkit end
+
 public class PrimaryLevelData implements ServerLevelData, WorldData {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -77,6 +90,20 @@
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
     private final TimerQueue<MinecraftServer> scheduledEvents;
+    // CraftBukkit start - Add world and pdc
+    public Registry<LevelStem> customDimensions;
+    private ServerLevel world;
+    protected Tag pdc;
+
+    public void setWorld(ServerLevel world) {
+        if (this.world != null) {
+            return;
+        }
+        this.world = world;
+        world.getWorld().readBukkitValues(pdc);
+        pdc = null;
+    }
+    // CraftBukkit end
 
     private PrimaryLevelData(@Nullable CompoundTag loadedPlayerTag, boolean wasModded, LevelData.RespawnData respawnData, long gameTime, long dayTime, int version, int clearWeatherTime, int rainTime, boolean raining, int thunderTime, boolean thundering, boolean initialized, boolean difficultyLocked, Optional<WorldBorder.Settings> legacyWorldBorderSettings, int wanderingTraderSpawnDelay, int wanderingTraderSpawnChance, @Nullable UUID wanderingTraderId, Set<String> knownServerBrands, Set<String> removedFeatureFlags, TimerQueue<MinecraftServer> scheduledEvents, @Nullable CompoundTag customBossEvents, EndDragonFight.Data endDragonFightData, LevelSettings settings, WorldOptions worldOptions, PrimaryLevelData.SpecialWorldProperty specialWorldProperty, Lifecycle worldGenSettingsLifecycle) {
         this.wasModded = wasModded;
@@ -113,7 +140,7 @@
 
     public static <T> PrimaryLevelData parse(Dynamic<T> input, LevelSettings settings, PrimaryLevelData.SpecialWorldProperty specialWorldProperty, WorldOptions worldOptions, Lifecycle worldGenSettingsLifecycle) {
         long i = input.get("Time").asLong(0L);
-        OptionalDynamic optionaldynamic = input.get("Player");
+        OptionalDynamic<T> optionaldynamic = input.get("Player"); // CraftBukkit - decompile error
         Codec codec = CompoundTag.CODEC;
 
         Objects.requireNonNull(codec);
@@ -132,7 +159,7 @@
         Optional optional = WorldBorder.Settings.CODEC.parse(input.get("world_border").orElseEmptyMap()).result();
         int k1 = input.get("WanderingTraderSpawnDelay").asInt(0);
         int l1 = input.get("WanderingTraderSpawnChance").asInt(0);
-        UUID uuid = (UUID) input.get("WanderingTraderId").read(UUIDUtil.CODEC).result().orElse((Object) null);
+        UUID uuid = (UUID) input.get("WanderingTraderId").read(UUIDUtil.CODEC).result().orElse(null); // CraftBukkit - decompile error
         Set set = (Set) input.get("ServerBrands").asStream().flatMap((dynamic1) -> {
             return dynamic1.asString().result().stream();
         }).collect(Collectors.toCollection(Sets::newLinkedHashSet));
@@ -141,7 +168,7 @@
         }).collect(Collectors.toSet());
         TimerQueue timerqueue = new TimerQueue(TimerCallbacks.SERVER_CALLBACKS, input.get("ScheduledEvents").asStream());
         CompoundTag compoundtag1 = (CompoundTag) input.get("CustomBossEvents").orElseEmptyMap().getValue();
-        DataResult dataresult = input.get("DragonFight").read(EndDragonFight.Data.CODEC);
+        DataResult<EndDragonFight.Data> dataresult = input.get("DragonFight").read(EndDragonFight.Data.CODEC); // CraftBukkit - decompile error
         Logger logger = PrimaryLevelData.LOGGER;
 
         Objects.requireNonNull(logger);
@@ -176,7 +203,7 @@
         tag.put("Version", compoundtag2);
         NbtUtils.addCurrentDataVersion(tag);
         DynamicOps<Tag> dynamicops = registryAccess.<Tag>createSerializationContext(NbtOps.INSTANCE);
-        DataResult dataresult = WorldGenSettings.encode(dynamicops, this.worldOptions, registryAccess);
+        DataResult<Tag> dataresult = WorldGenSettings.encode(dynamicops, this.worldOptions, new WorldDimensions(this.customDimensions != null ? this.customDimensions : registryAccess.lookupOrThrow(Registries.LEVEL_STEM))); // CraftBukkit
         Logger logger = PrimaryLevelData.LOGGER;
 
         Objects.requireNonNull(logger);
@@ -218,11 +245,13 @@
         tag.putInt("WanderingTraderSpawnDelay", this.wanderingTraderSpawnDelay);
         tag.putInt("WanderingTraderSpawnChance", this.wanderingTraderSpawnChance);
         tag.storeNullable("WanderingTraderId", UUIDUtil.CODEC, this.wanderingTraderId);
+        tag.putString("Bukkit.Version", Bukkit.getName() + "/" + Bukkit.getVersion() + "/" + Bukkit.getBukkitVersion()); // CraftBukkit
+        world.getWorld().storeBukkitValues(tag); // CraftBukkit - add pdc
     }
 
     private static ListTag stringCollectionToTag(Set<String> values) {
         ListTag listtag = new ListTag();
-        Stream stream = values.stream().map(StringTag::valueOf);
+        Stream<StringTag> stream = values.stream().map(StringTag::valueOf); // CraftBukkit - decompile error
 
         Objects.requireNonNull(listtag);
         stream.forEach(listtag::add);
@@ -291,6 +320,20 @@
 
     @Override
     public void setThundering(boolean thundering) {
+        // CraftBukkit start
+        if (this.thundering == thundering) {
+            return;
+        }
+
+        org.bukkit.World world = Bukkit.getWorld(getLevelName());
+        if (world != null) {
+            ThunderChangeEvent thunder = new ThunderChangeEvent(world, thundering);
+            Bukkit.getServer().getPluginManager().callEvent(thunder);
+            if (thunder.isCancelled()) {
+                return;
+            }
+        }
+        // CraftBukkit end
         this.thundering = thundering;
     }
 
@@ -311,6 +354,20 @@
 
     @Override
     public void setRaining(boolean raining) {
+        // CraftBukkit start
+        if (this.raining == raining) {
+            return;
+        }
+
+        org.bukkit.World world = Bukkit.getWorld(getLevelName());
+        if (world != null) {
+            WeatherChangeEvent weather = new WeatherChangeEvent(world, raining);
+            Bukkit.getServer().getPluginManager().callEvent(weather);
+            if (weather.isCancelled()) {
+                return;
+            }
+        }
+        // CraftBukkit end
         this.raining = raining;
     }
 
@@ -377,6 +434,12 @@
     @Override
     public void setDifficulty(Difficulty difficulty) {
         this.settings = this.settings.withDifficulty(difficulty);
+        // CraftBukkit start
+        ClientboundChangeDifficultyPacket packet = new ClientboundChangeDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+        for (ServerPlayer player : (java.util.List<ServerPlayer>) (java.util.List) world.players()) {
+            player.connection.send(packet);
+        }
+        // CraftBukkit end
     }
 
     @Override
@@ -511,6 +574,14 @@
         return this.settings.copy();
     }
 
+    // CraftBukkit start - Check if the name stored in NBT is the correct one
+    public void checkName(String name) {
+        if (!this.settings.levelName.equals(name)) {
+            this.settings.levelName = name;
+        }
+    }
+    // CraftBukkit end
+
     /** @deprecated */
     @Deprecated
     public static enum SpecialWorldProperty {
