--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -56,6 +56,10 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityDamageEvent;
+// CraftBukkit end
+
 public class Camel extends AbstractHorse {
 
     public static final float BABY_SCALE = 0.45F;
@@ -154,7 +158,7 @@
         ProfilerFiller profilerfiller = Profiler.get();
 
         profilerfiller.push("camelBrain");
-        Brain<?> brain = this.getBrain();
+        Brain<Camel> brain = (Brain<Camel>) this.getBrain(); // CraftBukkit - decompile error
 
         brain.tick(level, this);
         profilerfiller.pop();
@@ -476,9 +480,15 @@
     }
 
     @Override
-    protected void actuallyHurt(ServerLevel level, DamageSource source, float dmg) {
+    // CraftBukkit start - void -> boolean
+    public boolean actuallyHurt(ServerLevel level, DamageSource source, float dmg, EntityDamageEvent event) {
+        boolean damageResult = super.actuallyHurt(level, source, dmg, event);
+        if (!damageResult) {
+            return false;
+        }
+        // CraftBukkit end
         this.standUpInstantly();
-        super.actuallyHurt(level, source, dmg);
+        return true; // CraftBukkit
     }
 
     @Override
