--- a/net/minecraft/world/inventory/AbstractCraftingMenu.java
+++ b/net/minecraft/world/inventory/AbstractCraftingMenu.java
@@ -12,14 +_,16 @@
 public abstract class AbstractCraftingMenu extends RecipeBookMenu {
     private final int width;
     private final int height;
-    protected final CraftingContainer craftSlots;
-    protected final ResultContainer resultSlots = new ResultContainer();
+    public final CraftingContainer craftSlots;
+    public final ResultContainer resultSlots = new ResultContainer();
 
-    public AbstractCraftingMenu(MenuType<?> p_362493_, int p_360673_, int p_364200_, int p_363034_) {
+    public AbstractCraftingMenu(MenuType<?> p_362493_, int p_360673_, int p_364200_, int p_363034_, Inventory playerInventory) { // CraftBukkit
         super(p_362493_, p_360673_);
         this.width = p_364200_;
         this.height = p_363034_;
         this.craftSlots = new TransientCraftingContainer(this, p_364200_, p_363034_);
+        ((TransientCraftingContainer)this.craftSlots).resultInventory = this.resultSlots; // CraftBukkit - let InventoryCrafting know about its result slot
+        ((TransientCraftingContainer)this.craftSlots).setOwner(playerInventory.player);
     }
 
     protected Slot addResultSlot(Player p_362550_, int p_361054_, int p_363126_) {
