--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -47,7 +_,7 @@
     public static int ticksToBeFrog = Math.abs(-24000);
     public static final float HITBOX_WIDTH = 0.4F;
     public static final float HITBOX_HEIGHT = 0.3F;
-    private int age = 0;
+    public int age = 0;
     protected static final ImmutableList<SensorType<? extends Sensor<? super Tadpole>>> SENSOR_TYPES = ImmutableList.of(
         SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_PLAYERS, SensorType.HURT_BY, SensorType.FROG_TEMPTATIONS
     );
@@ -219,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_454577_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_454577_);
                 p_454577_.finalizeSpawn(serverlevel, serverlevel.getCurrentDifficultyAt(p_454577_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_454577_.setPersistenceRequired();
                 p_454577_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
