--- a/net/minecraft/world/entity/AgeableMob.java
+++ b/net/minecraft/world/entity/AgeableMob.java
@@ -24,6 +24,7 @@
     protected int age = 0;
     protected int forcedAge = 0;
     protected int forcedAgeTimer;
+    public boolean ageLocked; // CraftBukkit
 
     protected AgeableMob(EntityType<? extends AgeableMob> type, Level level) {
         super(type, level);
@@ -106,6 +107,7 @@
         super.addAdditionalSaveData(output);
         output.putInt("Age", this.getAge());
         output.putInt("ForcedAge", this.forcedAge);
+        output.putBoolean("AgeLocked", this.ageLocked); // CraftBukkit
     }
 
     @Override
@@ -113,6 +115,7 @@
         super.readAdditionalSaveData(input);
         this.setAge(input.getIntOr("Age", 0));
         this.forcedAge = input.getIntOr("ForcedAge", 0);
+        this.ageLocked = input.getBooleanOr("AgeLocked", this.ageLocked); // CraftBukkit
     }
 
     @Override
@@ -127,7 +130,7 @@
     @Override
     public void aiStep() {
         super.aiStep();
-        if (this.level().isClientSide()) {
+        if (this.level().isClientSide() || ageLocked) { // CraftBukkit
             if (this.forcedAgeTimer > 0) {
                 if (this.forcedAgeTimer % 4 == 0) {
                     this.level().addParticle(ParticleTypes.HAPPY_VILLAGER, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), 0.0D, 0.0D, 0.0D);
