--- a/net/minecraft/world/entity/EntityLightning.java
+++ b/net/minecraft/world/entity/EntityLightning.java
@@ -30,6 +30,11 @@
 import net.minecraft.world.phys.Vec3D;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityLightning extends Entity {
 
     private static final int START_LIFE = 2;
@@ -112,7 +117,7 @@
                     }
                 }
 
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             } else if (this.life < -this.random.nextInt(10)) {
                 --this.flashes;
                 this.life = 1;
@@ -121,7 +126,7 @@
             }
         }
 
-        if (this.life >= 0) {
+        if (this.life >= 0 && !this.visualOnly) { // CraftBukkit - add !this.visualOnly
             if (!(this.level() instanceof WorldServer)) {
                 this.level().setSkyFlashTime(2);
             } else if (!this.visualOnly) {
@@ -161,8 +166,12 @@
                 IBlockData iblockdata = BlockFireAbstract.getState(worldserver, blockposition);
 
                 if (worldserver.getBlockState(blockposition).isAir() && iblockdata.canSurvive(worldserver, blockposition)) {
-                    worldserver.setBlockAndUpdate(blockposition, iblockdata);
-                    ++this.blocksSetOnFire;
+                    // CraftBukkit start - add "!visualOnly"
+                    if (!visualOnly && !CraftEventFactory.callBlockIgniteEvent(worldserver, blockposition, this).isCancelled()) {
+                        worldserver.setBlockAndUpdate(blockposition, iblockdata);
+                        ++this.blocksSetOnFire;
+                    }
+                    // CraftBukkit end
                 }
 
                 for (int j = 0; j < i; ++j) {
@@ -170,8 +179,12 @@
 
                     iblockdata = BlockFireAbstract.getState(worldserver, blockposition1);
                     if (worldserver.getBlockState(blockposition1).isAir() && iblockdata.canSurvive(worldserver, blockposition1)) {
-                        worldserver.setBlockAndUpdate(blockposition1, iblockdata);
-                        ++this.blocksSetOnFire;
+                        // CraftBukkit start - add "!visualOnly"
+                        if (!visualOnly && !CraftEventFactory.callBlockIgniteEvent(worldserver, blockposition1, this).isCancelled()) {
+                            worldserver.setBlockAndUpdate(blockposition1, iblockdata);
+                            ++this.blocksSetOnFire;
+                        }
+                        // CraftBukkit end
                     }
                 }
 
