--- a/net/minecraft/world/level/chunk/EmptyLevelChunk.java
+++ b/net/minecraft/world/level/chunk/EmptyLevelChunk.java
@@ -1,5 +_,8 @@
 package net.minecraft.world.level.chunk;
 
+import ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk;
+import ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray;
+import ca.spottedleaf.moonrise.patches.starlight.light.StarLightEngine;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -13,7 +_,7 @@
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.material.Fluids;
 
-public class EmptyLevelChunk extends LevelChunk {
+public class EmptyLevelChunk extends LevelChunk implements StarlightChunk {
     private final Holder<Biome> biome;
 
     public EmptyLevelChunk(Level p_204422_, ChunkPos p_204423_, Holder<Biome> p_204424_) {
@@ -26,6 +_,13 @@
         return Blocks.VOID_AIR.defaultBlockState();
     }
 
+    // Paper start
+    @Override
+    public BlockState getBlockState(final int x, final int y, final int z) {
+        return Blocks.VOID_AIR.defaultBlockState();
+    }
+    // Paper end
+
     @Nullable
     @Override
     public BlockState setBlockState(BlockPos p_62605_, BlockState p_62606_, boolean p_62607_) {
@@ -84,4 +_,36 @@
     public Holder<Biome> getNoiseBiome(int p_204426_, int p_204427_, int p_204428_) {
         return this.biome;
     }
+
+    @Override
+    public SWMRNibbleArray[] starlight$getBlockNibbles() {
+        return StarLightEngine.getFilledEmptyLight(this.getLevel());
+    }
+
+    @Override
+    public void starlight$setBlockNibbles(final SWMRNibbleArray[] nibbles) {}
+
+    @Override
+    public SWMRNibbleArray[] starlight$getSkyNibbles() {
+        return StarLightEngine.getFilledEmptyLight(this.getLevel());
+    }
+
+    @Override
+    public void starlight$setSkyNibbles(final SWMRNibbleArray[] nibbles) {}
+
+    @Override
+    public boolean[] starlight$getSkyEmptinessMap() {
+        return null;
+    }
+
+    @Override
+    public void starlight$setSkyEmptinessMap(final boolean[] emptinessMap) {}
+
+    @Override
+    public boolean[] starlight$getBlockEmptinessMap() {
+        return null;
+    }
+
+    @Override
+    public void starlight$setBlockEmptinessMap(final boolean[] emptinessMap) {}
 }
