--- a/net/minecraft/world/level/dimension/end/DragonRespawnAnimation.java
+++ b/net/minecraft/world/level/dimension/end/DragonRespawnAnimation.java
@@ -12,6 +12,10 @@
 import net.minecraft.world.level.levelgen.feature.SpikeFeature;
 import net.minecraft.world.level.levelgen.feature.configurations.SpikeConfiguration;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public enum DragonRespawnAnimation {
 
     START {
@@ -23,7 +27,7 @@
                 endcrystal.setBeamTarget(blockpos1);
             }
 
-            fight.setRespawnStage(null.PREPARING_TO_SUMMON_PILLARS);
+            fight.setRespawnStage(PREPARING_TO_SUMMON_PILLARS); // CraftBukkit - decompile error
         }
     },
     PREPARING_TO_SUMMON_PILLARS {
@@ -34,7 +38,7 @@
                     level.levelEvent(3001, new BlockPos(0, 128, 0), 0);
                 }
             } else {
-                fight.setRespawnStage(null.SUMMONING_PILLARS);
+                fight.setRespawnStage(SUMMONING_PILLARS); // CraftBukkit - decompile error
             }
 
         }
@@ -70,7 +74,7 @@
                         Feature.END_SPIKE.place(spikeconfiguration, level, level.getChunkSource().getGenerator(), RandomSource.create(), new BlockPos(spikefeature_endspike.getCenterX(), 45, spikefeature_endspike.getCenterZ()));
                     }
                 } else if (flag) {
-                    fight.setRespawnStage(null.SUMMONING_DRAGON);
+                    fight.setRespawnStage(SUMMONING_DRAGON); // CraftBukkit - decompile error
                 }
             }
 
@@ -80,13 +84,13 @@
         @Override
         public void tick(ServerLevel level, EndDragonFight fight, List<EndCrystal> crystals, int time, BlockPos portal) {
             if (time >= 100) {
-                fight.setRespawnStage(null.END);
+                fight.setRespawnStage(END); // CraftBukkit - decompile error
                 fight.resetSpikeCrystals();
 
                 for (EndCrystal endcrystal : crystals) {
                     endcrystal.setBeamTarget((BlockPos) null);
                     level.explode(endcrystal, endcrystal.getX(), endcrystal.getY(), endcrystal.getZ(), 6.0F, Level.ExplosionInteraction.NONE);
-                    endcrystal.discard();
+                    endcrystal.discard(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
                 }
             } else if (time >= 80) {
                 level.levelEvent(3001, new BlockPos(0, 128, 0), 0);
