--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -105,14 +105,14 @@
         this.origin = origin;
         this.validPlayer = EntitySelector.ENTITY_STILL_ALIVE.and(EntitySelector.withinDistance((double) origin.getX(), (double) (128 + origin.getY()), (double) origin.getZ(), 192.0D));
         this.needsStateScanning = dragonFightData.needsStateScanning;
-        this.dragonUUID = (UUID) dragonFightData.dragonUUID.orElse((Object) null);
+        this.dragonUUID = (UUID) dragonFightData.dragonUUID.orElse(null); // CraftBukkit - decompile error
         this.dragonKilled = dragonFightData.dragonKilled;
         this.previouslyKilled = dragonFightData.previouslyKilled;
         if (dragonFightData.isRespawning) {
             this.respawnStage = DragonRespawnAnimation.START;
         }
 
-        this.portalLocation = (BlockPos) dragonFightData.exitPortalLocation.orElse((Object) null);
+        this.portalLocation = (BlockPos) dragonFightData.exitPortalLocation.orElse(null); // CraftBukkit - decompile error
         this.gateways.addAll((Collection) dragonFightData.gateways.orElseGet(() -> {
             ObjectArrayList<Integer> objectarraylist = new ObjectArrayList(ContiguousSet.create(Range.closedOpen(0, 20), DiscreteDomain.integers()));
 
@@ -202,7 +202,7 @@
             this.dragonKilled = false;
             if (!flag) {
                 EndDragonFight.LOGGER.info("But we didn't have a portal, let's remove it.");
-                enderdragon.discard();
+                enderdragon.discard(null); // CraftBukkit - add Bukkit remove cause
                 this.dragonUUID = null;
             }
         }
@@ -478,7 +478,7 @@
         return this.previouslyKilled;
     }
 
-    public void tryRespawn() {
+    public boolean tryRespawn() { // CraftBukkit - return boolean
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos blockpos = this.portalLocation;
 
@@ -503,19 +503,19 @@
                 List<EndCrystal> list1 = this.level.<EndCrystal>getEntitiesOfClass(EndCrystal.class, new AABB(blockpos1.relative(direction, 2)));
 
                 if (list1.isEmpty()) {
-                    return;
+                    return false; // CraftBukkit - return value
                 }
 
                 list.addAll(list1);
             }
 
             EndDragonFight.LOGGER.debug("Found all crystals, respawning dragon.");
-            this.respawnDragon(list);
+            return this.respawnDragon(list); // CraftBukkit - return value
         }
-
+        return false; // CraftBukkit - return value
     }
 
-    public void respawnDragon(List<EndCrystal> crystals) {
+    public boolean respawnDragon(List<EndCrystal> crystals) { // CraftBukkit - return boolean
         if (this.dragonKilled && this.respawnStage == null) {
             for (BlockPattern.BlockPatternMatch blockpattern_blockpatternmatch = this.findExitPortal(); blockpattern_blockpatternmatch != null; blockpattern_blockpatternmatch = this.findExitPortal()) {
                 for (int i = 0; i < this.exitPortalPattern.getWidth(); ++i) {
@@ -535,8 +535,9 @@
             this.respawnTime = 0;
             this.spawnExitPortal(false);
             this.respawnCrystals = crystals;
+            return true; // CraftBukkit - return value
         }
-
+        return false; // CraftBukkit - return value
     }
 
     public void resetSpikeCrystals() {
