--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -71,13 +_,14 @@
     protected int inGroundTime;
     public AbstractArrow.Pickup pickup = AbstractArrow.Pickup.DISALLOWED;
     public int shakeTime = 0;
-    private int life = 0;
-    private double baseDamage = 2.0;
+    public int life = 0;
+    public double baseDamage = 2.0;
     private SoundEvent soundEvent = this.getDefaultHitGroundSoundEvent();
     private @Nullable IntOpenHashSet piercingIgnoreEntityIds;
     private @Nullable List<Entity> piercedAndKilledEntities;
-    private ItemStack pickupItemStack = this.getDefaultPickupItem();
-    private @Nullable ItemStack firedFromWeapon = null;
+    public ItemStack pickupItemStack = this.getDefaultPickupItem();
+    @Nullable
+    public ItemStack firedFromWeapon = null;
 
     protected AbstractArrow(EntityType<? extends AbstractArrow> p_479294_, Level p_479811_) {
         super(p_479294_, p_479811_);
@@ -191,7 +_,8 @@
             this.shakeTime--;
         }
 
-        if (this.isInWaterOrRain()) {
+        // Neo: Allow custom fluids to extinguish flaming arrows
+        if (this.isInWaterOrRain() || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType))) {
             this.clearFire();
         }
 
@@ -282,11 +_,15 @@
 
             if (arraylist.isEmpty()) {
                 if (this.isAlive() && p_479463_.getType() != HitResult.Type.MISS) {
+                    if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, p_479463_))
+                        break;
                     this.hitTargetOrDeflectSelf(p_479463_);
                     this.needsSync = true;
                 }
                 break;
-            } else if (this.isAlive() && !this.noPhysics) {
+            } else if (this.isAlive() && !this.noPhysics && entityhitresult.getType() != HitResult.Type.MISS) {
+                if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, entityhitresult))
+                    break;
                 ProjectileDeflection projectiledeflection = this.hitTargetsOrDeflectSelf(arraylist);
                 this.needsSync = true;
                 if (this.getPierceLevel() > 0 && projectiledeflection == ProjectileDeflection.NONE) {
@@ -341,7 +_,7 @@
         this.life = 0;
     }
 
-    protected boolean isInGround() {
+    public boolean isInGround() {
         return this.entityData.get(IN_GROUND);
     }
 
@@ -675,7 +_,7 @@
         this.setFlag(1, p_480262_);
     }
 
-    private void setPierceLevel(byte p_479996_) {
+    public void setPierceLevel(byte p_479996_) {
         this.entityData.set(PIERCE_LEVEL, p_479996_);
     }
 
