--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -15,6 +_,7 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.UUID;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -39,6 +_,8 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.saveddata.SavedData;
 import net.minecraft.world.level.saveddata.SavedDataType;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.map.CraftMapView;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -64,21 +_,28 @@
             )
             .apply(p_400959_, MapItemSavedData::new)
     );
-    public final int centerX;
-    public final int centerZ;
-    public final ResourceKey<Level> dimension;
-    private final boolean trackingPosition;
-    private final boolean unlimitedTracking;
-    public final byte scale;
+    public int centerX;
+    public int centerZ;
+    public ResourceKey<Level> dimension;
+    public boolean trackingPosition;
+    public boolean unlimitedTracking;
+    public byte scale;
     public byte[] colors = new byte[16384];
-    public final boolean locked;
+    public boolean locked;
     private final List<MapItemSavedData.HoldingPlayer> carriedBy = Lists.newArrayList();
     private final Map<Player, MapItemSavedData.HoldingPlayer> carriedByPlayers = Maps.newHashMap();
     private final Map<String, MapBanner> bannerMarkers = Maps.newHashMap();
-    final Map<String, MapDecoration> decorations = Maps.newLinkedHashMap();
+    public final Map<String, MapDecoration> decorations = Maps.newLinkedHashMap();
     private final Map<String, MapFrame> frameMarkers = Maps.newHashMap();
     private int trackedDecorationCount;
 
+    // CraftBukkit start
+    public final CraftMapView mapView;
+    private CraftServer server;
+    public UUID uniqueId = null;
+    public MapId id;
+    // CraftBukkit end
+
     public static SavedDataType<MapItemSavedData> type(MapId p_401062_) {
         return new SavedDataType<>(p_401062_.key(), () -> {
             throw new IllegalStateException("Should never create an empty map saved data");
@@ -95,6 +_,10 @@
         this.trackingPosition = p_164771_;
         this.unlimitedTracking = p_164772_;
         this.locked = p_164773_;
+        // CraftBukkit start
+        mapView = new CraftMapView(this);
+        server = (CraftServer) org.bukkit.Bukkit.getServer();
+        // CraftBukkit end
     }
 
     private MapItemSavedData(
@@ -363,7 +_,7 @@
         return mapitemsaveddata$holdingplayer == null ? null : mapitemsaveddata$holdingplayer.nextUpdatePacket(p_323760_);
     }
 
-    private void setColorsDirty(int p_164790_, int p_164791_) {
+    public void setColorsDirty(int p_164790_, int p_164791_) {
         this.setDirty();
 
         for (MapItemSavedData.HoldingPlayer mapitemsaveddata$holdingplayer : this.carriedBy) {
