--- a/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
+++ b/net/minecraft/world/scores/criteria/ObjectiveCriteria.java
@@ -15,7 +_,7 @@
 
 public class ObjectiveCriteria {
     private static final Map<String, ObjectiveCriteria> CUSTOM_CRITERIA = Maps.newHashMap();
-    private static final Map<String, ObjectiveCriteria> CRITERIA_CACHE = Maps.newHashMap();
+    public static final Map<String, ObjectiveCriteria> CRITERIA_CACHE = Maps.newHashMap();
     public static final Codec<ObjectiveCriteria> CODEC = Codec.STRING
         .comapFlatMap(
             p_401000_ -> byName(p_401000_).map(DataResult::success).orElse(DataResult.error(() -> "No scoreboard criteria with name: " + p_401000_)),
