--- a/net/minecraft/world/food/FoodProperties.java
+++ b/net/minecraft/world/food/FoodProperties.java
@@ -44,7 +_,7 @@
             null, p_366505_.getX(), p_366505_.getY(), p_366505_.getZ(), p_366719_.sound().value(), SoundSource.NEUTRAL, 1.0F, randomsource.triangle(1.0F, 0.4F)
         );
         if (p_366505_ instanceof Player player) {
-            player.getFoodData().eat(this);
+            player.getFoodData().eat(this, p_366556_, (net.minecraft.server.level.ServerPlayer) player); // CraftBukkit
             p_366676_.playSound(
                 null,
                 player.getX(),
