--- a/net/minecraft/world/entity/animal/equine/Horse.java
+++ b/net/minecraft/world/entity/animal/equine/Horse.java
@@ -83,7 +_,7 @@
         return this.entityData.get(DATA_ID_TYPE_VARIANT);
     }
 
-    private void setVariantAndMarkings(Variant p_481047_, Markings p_478007_) {
+    public void setVariantAndMarkings(Variant p_481047_, Markings p_478007_) {
         this.setTypeVariant(p_481047_.getId() & 0xFF | p_478007_.getId() << 8 & 0xFF00);
     }
 
