--- a/net/minecraft/network/chat/Component.java
+++ b/net/minecraft/network/chat/Component.java
@@ -25,7 +_,7 @@
 import net.minecraft.world.level.ChunkPos;
 import org.jspecify.annotations.Nullable;
 
-public interface Component extends Message, FormattedText {
+public interface Component extends Message, FormattedText, Iterable<Component> { // CraftBukkit
     Style getStyle();
 
     ComponentContents getContents();
