--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -30,12 +_,20 @@
 
         for (short short1 : p_285027_) {
             this.positions[j] = short1;
-            this.states[j] = p_285414_.getBlockState(
+            this.states[j] = (p_285414_ != null) ? p_285414_.getBlockState(
                 SectionPos.sectionRelativeX(short1), SectionPos.sectionRelativeY(short1), SectionPos.sectionRelativeZ(short1)
-            );
+            ) : net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
             j++;
         }
     }
+
+    // CraftBukkit start - Add constructor
+    public ClientboundSectionBlocksUpdatePacket(SectionPos sectionpos, ShortSet shortset, BlockState[] states) {
+        this.sectionPos = sectionpos;
+        this.positions = shortset.toShortArray();
+        this.states = states;
+    }
+    // CraftBukkit end
 
     private ClientboundSectionBlocksUpdatePacket(FriendlyByteBuf p_179196_) {
         this.sectionPos = SectionPos.STREAM_CODEC.decode(p_179196_);
