--- a/net/minecraft/world/entity/animal/pig/EntityPig.java
+++ b/net/minecraft/world/entity/animal/pig/EntityPig.java
@@ -55,6 +55,11 @@
 import net.minecraft.world.phys.Vec3D;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityPig extends EntityAnimal implements ISteerable {
 
     private static final DataWatcherObject<Integer> DATA_BOOST_TIME = DataWatcher.<Integer>defineId(EntityPig.class, DataWatcherRegistry.INT);
@@ -196,7 +201,14 @@
             EntityPigZombie entitypigzombie = (EntityPigZombie) this.convertTo(EntityTypes.ZOMBIFIED_PIGLIN, ConversionParams.single(this, false, true), (entitypigzombie1) -> {
                 entitypigzombie1.populateDefaultEquipmentSlots(this.getRandom(), worldserver.getCurrentDifficultyAt(this.blockPosition()));
                 entitypigzombie1.setPersistenceRequired();
-            });
+            // CraftBukkit start
+            }, null, null);
+            if (CraftEventFactory.callPigZapEvent(this, entitylightning, entitypigzombie).isCancelled()) {
+                return;
+            }
+            worldserver.addFreshEntity(entitypigzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+            this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
+            // CraftBukkit end
 
             if (entitypigzombie == null) {
                 super.thunderHit(worldserver, entitylightning);
