--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -45,7 +_,7 @@
     );
     public static final SavedDataType<Raids> TYPE = new SavedDataType<>("raids", Raids::new, CODEC, DataFixTypes.SAVED_DATA_RAIDS);
     public static final SavedDataType<Raids> TYPE_END = new SavedDataType<>("raids_end", Raids::new, CODEC, DataFixTypes.SAVED_DATA_RAIDS);
-    private final Int2ObjectMap<Raid> raidMap = new Int2ObjectOpenHashMap<>();
+    public final Int2ObjectMap<Raid> raidMap = new Int2ObjectOpenHashMap<>();
     private int nextId = 1;
     private int tick;
 
