--- a/net/minecraft/server/players/NameAndId.java
+++ b/net/minecraft/server/players/NameAndId.java
@@ -50,4 +_,10 @@
         UUID uuid = UUIDUtil.createOfflinePlayerUUID(p_433923_);
         return new NameAndId(uuid, p_433923_);
     }
+
+    // Paper start - utility method for common conversion back to the game profile
+    public com.mojang.authlib.GameProfile toUncompletedGameProfile() {
+        return new com.mojang.authlib.GameProfile(this.id, this.name);
+    }
+    // Paper end - utility method for common conversion back to the game profile
 }
