--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -66,6 +_,7 @@
 import net.minecraft.server.PlayerAdvancements;
 import net.minecraft.server.RegistryLayer;
 import net.minecraft.server.ServerScoreboard;
+import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.CommonListenerCookie;
@@ -99,6 +_,9 @@
 import net.minecraft.world.scores.Objective;
 import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.Team;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.event.player.PlayerRespawnEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -113,7 +_,7 @@
     private static final int SEND_PLAYER_INFO_INTERVAL = 600;
     private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT);
     private final MinecraftServer server;
-    private final List<ServerPlayer> players = Lists.newArrayList();
+    public final List<ServerPlayer> players = Lists.newArrayList();
     private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
     private final UserBanList bans;
     private final IpBanList ipBans;
@@ -121,14 +_,22 @@
     private final UserWhiteList whitelist;
     private final Map<UUID, ServerStatsCounter> stats = Maps.newHashMap();
     private final Map<UUID, PlayerAdvancements> advancements = Maps.newHashMap();
-    private final PlayerDataStorage playerIo;
+    public final PlayerDataStorage playerIo;
     private final LayeredRegistryAccess<RegistryLayer> registries;
     private int viewDistance;
     private int simulationDistance;
     private boolean allowCommandsForAllPlayers;
     private int sendAllPlayerInfoIn;
+    private final List<ServerPlayer> playersView = java.util.Collections.unmodifiableList(players);
+
+    // CraftBukkit start
+    private CraftServer cserver;
+    private final Map<String,ServerPlayer> playersByName = new java.util.HashMap<>();
 
     public PlayerList(MinecraftServer p_203842_, LayeredRegistryAccess<RegistryLayer> p_251844_, PlayerDataStorage p_203844_, NotificationService p_443409_) {
+        this.cserver = p_203842_.server = new CraftServer((DedicatedServer) p_203842_, this);
+        p_203842_.console = org.bukkit.craftbukkit.command.ColouredConsoleSender.getInstance();
+        // CraftBukkit end
         this.server = p_203842_;
         this.registries = p_251844_;
         this.playerIo = p_203844_;
@@ -158,7 +_,7 @@
         LevelData leveldata = serverlevel.getLevelData();
         ServerGamePacketListenerImpl servergamepacketlistenerimpl = new ServerGamePacketListenerImpl(this.server, p_11262_, p_11263_, p_301988_);
         p_11262_.setupInboundProtocol(
-            GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()), servergamepacketlistenerimpl),
+            GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess(), servergamepacketlistenerimpl.getConnectionType()), servergamepacketlistenerimpl),
             servergamepacketlistenerimpl
         );
         servergamepacketlistenerimpl.suspendFlushing();
@@ -184,10 +_,12 @@
         servergamepacketlistenerimpl.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
         servergamepacketlistenerimpl.send(new ClientboundPlayerAbilitiesPacket(p_11263_.getAbilities()));
         servergamepacketlistenerimpl.send(new ClientboundSetHeldSlotPacket(p_11263_.getInventory().getSelectedSlot()));
+        var dataSyncEvent = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.OnDatapackSyncEvent(this, p_11263_));
         RecipeManager recipemanager = this.server.getRecipeManager();
         servergamepacketlistenerimpl.send(
             new ClientboundUpdateRecipesPacket(recipemanager.getSynchronizedItemProperties(), recipemanager.getSynchronizedStonecutterRecipes())
         );
+        net.neoforged.neoforge.common.CommonHooks.sendRecipes(p_11263_, dataSyncEvent.getRecipeTypesToSend(), recipemanager.recipeMap());
         this.sendPlayerPermissionLevel(p_11263_);
         p_11263_.getStats().markAllDirty();
         p_11263_.getRecipeBook().sendInitialRecipeBook(p_11263_);
@@ -218,9 +_,12 @@
         p_11263_.initInventoryMenu();
         this.server.notificationManager().playerJoined(p_11263_);
         servergamepacketlistenerimpl.resumeFlushing();
+
+        net.neoforged.neoforge.attachment.AttachmentSync.syncInitialPlayerAttachments(p_11263_);
+        net.neoforged.neoforge.event.EventHooks.firePlayerLoggedIn(p_11263_);
     }
 
-    protected void updateEntireScoreboard(ServerScoreboard p_11274_, ServerPlayer p_11275_) {
+    public void updateEntireScoreboard(ServerScoreboard p_11274_, ServerPlayer p_11275_) {
         Set<Objective> set = Sets.newHashSet();
 
         for (PlayerTeam playerteam : p_11274_.getPlayerTeams()) {
@@ -287,6 +_,7 @@
     }
 
     protected void save(ServerPlayer p_11277_) {
+        if (p_11277_.connection == null) return;
         this.playerIo.save(p_11277_);
         ServerStatsCounter serverstatscounter = this.stats.get(p_11277_.getUUID());
         if (serverstatscounter != null) {
@@ -300,6 +_,7 @@
     }
 
     public void remove(ServerPlayer p_11287_) {
+        net.neoforged.neoforge.event.EventHooks.firePlayerLoggedOut(p_11287_);
         ServerLevel serverlevel = p_11287_.level();
         p_11287_.awardStat(Stats.LEAVE_GAME);
         this.save(p_11287_);
@@ -385,17 +_,32 @@
         return !set.isEmpty();
     }
 
+    public ServerPlayer respawn(ServerPlayer serverplayer, boolean flag, Entity.RemovalReason entity_removalreason, PlayerRespawnEvent.RespawnReason reason, Location location) {
+        return respawn(serverplayer, flag, entity_removalreason);
+    }
+
+    public ServerPlayer respawn(ServerPlayer serverplayer, boolean flag, Entity.RemovalReason entity_removalreason, PlayerRespawnEvent.RespawnReason reason) {
+        return respawn(serverplayer, flag, entity_removalreason, PlayerRespawnEvent.RespawnReason.PLUGIN, null);
+    }
+
     public ServerPlayer respawn(ServerPlayer p_11237_, boolean p_11238_, Entity.RemovalReason p_348558_) {
         TeleportTransition teleporttransition = p_11237_.findRespawnPositionAndUseSpawnBlock(!p_11238_, TeleportTransition.DO_NOTHING);
         this.players.remove(p_11237_);
         p_11237_.level().removePlayerImmediately(p_11237_, p_348558_);
+
+        // Neo: Allow changing the respawn position of players. The local dimension transition is updated with the new target.
+        var event = net.neoforged.neoforge.event.EventHooks.firePlayerRespawnPositionEvent(p_11237_, teleporttransition, p_11238_);
+        teleporttransition = event.getTeleportTransition();
+
         ServerLevel serverlevel = teleporttransition.newLevel();
         ServerPlayer serverplayer = new ServerPlayer(this.server, serverlevel, p_11237_.getGameProfile(), p_11237_.clientInformation());
         serverplayer.connection = p_11237_.connection;
         serverplayer.restoreFrom(p_11237_, p_11238_);
         serverplayer.setId(p_11237_.getId());
         serverplayer.setMainArm(p_11237_.getMainArm());
-        if (!teleporttransition.missingRespawnBlock()) {
+
+        // Neo: Allow the event to control if the original spawn position is copied
+        if (event.copyOriginalSpawnPosition()) {
             serverplayer.copyRespawnPosition(p_11237_);
         }
 
@@ -426,6 +_,8 @@
         this.playersByUUID.put(serverplayer.getUUID(), serverplayer);
         serverplayer.initInventoryMenu();
         serverplayer.setHealth(serverplayer.getHealth());
+        net.neoforged.neoforge.attachment.AttachmentSync.syncInitialPlayerAttachments(serverplayer);
+        net.neoforged.neoforge.event.EventHooks.firePlayerRespawnEvent(serverplayer, p_11238_);
         ServerPlayer.RespawnConfig serverplayer$respawnconfig = serverplayer.getRespawnConfig();
         if (!p_11238_ && serverplayer$respawnconfig != null) {
             LevelData.RespawnData leveldata$respawndata = serverplayer$respawnconfig.respawnData();
@@ -539,6 +_,7 @@
     }
 
     public void op(NameAndId p_442929_, Optional<LevelBasedPermissionSet> p_443490_, Optional<Boolean> p_443080_) {
+        if (net.neoforged.neoforge.event.EventHooks.onPermissionChanged(p_442929_, this.server.operatorUserPermissions(), this)) return;
         this.ops
             .add(
                 new ServerOpListEntry(
@@ -552,6 +_,7 @@
     }
 
     public void deop(NameAndId p_435771_) {
+        if (net.neoforged.neoforge.event.EventHooks.onPermissionChanged(p_435771_, LevelBasedPermissionSet.ALL, this)) return;
         if (this.ops.remove(p_435771_)) {
             ServerPlayer serverplayer = this.getPlayer(p_435771_.id());
             if (serverplayer != null) {
@@ -572,6 +_,7 @@
             p_11227_.connection.send(new ClientboundEntityEventPacket(p_11227_, b0));
         }
 
+        p_11227_.getBukkitEntity().recalculatePermissions(); // CraftBukkit
         this.server.getCommands().sendCommands(p_11227_);
     }
 
@@ -642,8 +_,12 @@
     public void sendLevelInfo(ServerPlayer p_11230_, ServerLevel p_11231_) {
         WorldBorder worldborder = p_11231_.getWorldBorder();
         p_11230_.connection.send(new ClientboundInitializeBorderPacket(worldborder));
+        if (p_11230_.connection.hasChannel(net.neoforged.neoforge.network.payload.ClientboundCustomSetTimePayload.TYPE)) {
+            p_11230_.connection.send(new net.neoforged.neoforge.network.payload.ClientboundCustomSetTimePayload(p_11231_.getGameTime(), p_11231_.getDayTime(), p_11231_.getGameRules().get(GameRules.ADVANCE_TIME), p_11231_.getDayTimeFraction(), p_11231_.getDayTimePerTick()));
+        } else {
         p_11230_.connection
             .send(new ClientboundSetTimePacket(p_11231_.getGameTime(), p_11231_.getDayTime(), p_11231_.getGameRules().get(GameRules.ADVANCE_TIME)));
+        }
         p_11230_.connection.send(new ClientboundSetDefaultSpawnPositionPacket(p_11231_.getRespawnData()));
         if (p_11231_.isRaining()) {
             p_11230_.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
@@ -653,6 +_,7 @@
 
         p_11230_.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.LEVEL_CHUNKS_LOAD_START, 0.0F));
         this.server.tickRateManager().updateJoiningPlayer(p_11230_);
+        net.neoforged.neoforge.attachment.AttachmentSync.syncInitialLevelAttachments(p_11231_, p_11230_);
     }
 
     public void sendAllPlayerInfo(ServerPlayer p_11293_) {
@@ -758,14 +_,28 @@
     }
 
     public ServerStatsCounter getPlayerStats(Player p_11240_) {
-        GameProfile gameprofile = p_11240_.getGameProfile();
-        return this.stats.computeIfAbsent(gameprofile.id(), p_460349_ -> {
+        // CraftBukkit start
+        return getPlayerStats(p_11240_.nameAndId());
+    }
+
+    public ServerStatsCounter getPlayerStats(ServerPlayer entityhuman) {
+        NameAndId gameprofile = entityhuman.nameAndId();
+        ServerStatsCounter serverstatisticmanager = entityhuman.getStats();
+        return serverstatisticmanager == null ? getPlayerStats(gameprofile) : serverstatisticmanager;
+    }
+
+    public ServerStatsCounter getPlayerStats(NameAndId gameprofile) {
+        ServerPlayer entityhuman = this.getPlayer(gameprofile.id());
+        ServerStatsCounter serverstatisticmanager = entityhuman == null ? null : entityhuman.getStats();
+        if (serverstatisticmanager == null) {
             Path path = this.locateStatsFile(gameprofile);
+
             return new ServerStatsCounter(this.server, path);
-        });
+        }
+        return serverstatisticmanager;
     }
 
-    private Path locateStatsFile(GameProfile p_460971_) {
+    private Path locateStatsFile(NameAndId p_460971_) {
         Path path = this.server.getWorldPath(LevelResource.PLAYER_STATS_DIR);
         Path path1 = path.resolve(p_460971_.id() + ".json");
         if (Files.exists(path1)) {
@@ -797,6 +_,8 @@
             this.advancements.put(uuid, playeradvancements);
         }
 
+        // Forge: don't overwrite active player with a fake one.
+        if (!(p_11297_ instanceof net.neoforged.neoforge.common.util.FakePlayer))
         playeradvancements.setPlayer(p_11297_);
         return playeradvancements;
     }
@@ -820,7 +_,7 @@
     }
 
     public List<ServerPlayer> getPlayers() {
-        return this.players;
+        return this.playersView; // Neo: Return an unmodifiable view, we don't want people removing things without us knowing.
     }
 
     public @Nullable ServerPlayer getPlayer(UUID p_11260_) {
@@ -846,6 +_,7 @@
             playeradvancements.reload(this.server.getAdvancements());
         }
 
+        var dataSyncEvent = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.OnDatapackSyncEvent(this, null));
         this.broadcastAll(new ClientboundUpdateTagsPacket(TagNetworkSerialization.serializeTagsToNetwork(this.registries)));
         RecipeManager recipemanager = this.server.getRecipeManager();
         ClientboundUpdateRecipesPacket clientboundupdaterecipespacket = new ClientboundUpdateRecipesPacket(
@@ -855,10 +_,27 @@
         for (ServerPlayer serverplayer : this.players) {
             serverplayer.connection.send(clientboundupdaterecipespacket);
             serverplayer.getRecipeBook().sendInitialRecipeBook(serverplayer);
-        }
+            net.neoforged.neoforge.common.CommonHooks.sendRecipes(serverplayer, dataSyncEvent.getRecipeTypesToSend(), recipemanager.recipeMap());
+        }
+    }
+
+    public void reloadRecipes() {
+        // CraftBukkit end
+        RecipeManager recipemanager = this.server.getRecipeManager();
+        ClientboundUpdateRecipesPacket clientboundupdaterecipespacket = new ClientboundUpdateRecipesPacket(recipemanager.getSynchronizedItemProperties(), recipemanager.getSynchronizedStonecutterRecipes());
+
+        for (ServerPlayer serverplayer : this.players) {
+            serverplayer.connection.send(clientboundupdaterecipespacket);
+            serverplayer.getRecipeBook().sendInitialRecipeBook(serverplayer);
+        }
+
     }
 
     public boolean isAllowCommandsForAllPlayers() {
         return this.allowCommandsForAllPlayers;
+    }
+
+    public PlayerDataStorage getPlayerIo() {
+        return this.playerIo;
     }
 }
