--- a/net/minecraft/world/entity/animal/cow/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/cow/MushroomCow.java
@@ -54,7 +_,8 @@
     private static final EntityDataAccessor<Integer> DATA_TYPE = SynchedEntityData.defineId(MushroomCow.class, EntityDataSerializers.INT);
     private static final int MUTATE_CHANCE = 1024;
     private static final String TAG_STEW_EFFECTS = "stew_effects";
-    private @Nullable SuspiciousStewEffects stewEffects;
+    @Nullable
+    public SuspiciousStewEffects stewEffects;
     private @Nullable UUID lastLightningBoltUUID;
 
     public MushroomCow(EntityType<? extends MushroomCow> p_478098_, Level p_479607_) {
@@ -171,11 +_,19 @@
     @Override
     public void shear(ServerLevel p_477948_, SoundSource p_480914_, ItemStack p_479429_) {
         p_477948_.playSound(null, this, SoundEvents.MOOSHROOM_SHEAR, p_480914_, 1.0F, 1.0F);
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.COW, (timer) -> {})) return;
         this.convertTo(EntityType.COW, ConversionParams.single(this, false, false), p_478543_ -> {
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_478543_);
             p_477948_.sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(0.5), this.getZ(), 1, 0.0, 0.0, 0.0, 0.0);
             this.dropFromShearingLootTable(p_477948_, BuiltInLootTables.SHEAR_MOOSHROOM, p_479429_, (p_478407_, p_481493_) -> {
                 for (int i = 0; i < p_481493_.getCount(); i++) {
-                    p_478407_.addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(1.0), this.getZ(), p_481493_.copyWithCount(1)));
+                    // Neo: Change from addFreshEntity to spawnAtLocation to ensure captureDrops can capture this, we also need to unset the default pickup delay from the item
+                    // Vanilla uses this.getY(1.0) for the y-level, which is this.getY() + this.getBbHeight() * 1.0, so we pass the BB height as the Y-offset.
+                    ItemEntity item = spawnAtLocation(p_477948_, p_481493_.copyWithCount(1), this.getBbHeight());
+                    if (item != null) {
+                        // addFreshEntity does not incur a pickup delay, while spawnAtLocation sets the default pickup delay.
+                        item.setNoPickUpDelay();
+                    }
                 }
             });
         });
@@ -205,7 +_,7 @@
         return suspiciouseffectholder != null ? Optional.of(suspiciouseffectholder.getSuspiciousEffects()) : Optional.empty();
     }
 
-    private void setVariant(MushroomCow.Variant p_479508_) {
+    public void setVariant(MushroomCow.Variant p_479508_) {
         this.entityData.set(DATA_TYPE, p_479508_.id);
     }
 
