--- a/net/minecraft/server/commands/CommandGamerule.java
+++ b/net/minecraft/server/commands/CommandGamerule.java
@@ -41,7 +41,7 @@
         CommandListenerWrapper commandlistenerwrapper = (CommandListenerWrapper) commandcontext.getSource();
         T t0 = (T) commandcontext.getArgument("value", gamerule.valueClass());
 
-        commandlistenerwrapper.getLevel().getGameRules().set(gamerule, t0, ((CommandListenerWrapper) commandcontext.getSource()).getServer());
+        commandlistenerwrapper.getLevel().getGameRules().set(gamerule, t0, ((CommandListenerWrapper) commandcontext.getSource()).getLevel()); // CraftBukkit
         commandlistenerwrapper.sendSuccess(() -> {
             return IChatBaseComponent.translatable("commands.gamerule.set", gamerule.id(), gamerule.serialize(t0));
         }, true);
