--- a/net/minecraft/server/commands/CommandSpawnpoint.java
+++ b/net/minecraft/server/commands/CommandSpawnpoint.java
@@ -43,7 +43,7 @@
         float f1 = MathHelper.clamp(vec2f.x, -90.0F, 90.0F);
 
         for (EntityPlayer entityplayer : collection) {
-            entityplayer.setRespawnPosition(new EntityPlayer.RespawnConfig(WorldData.a.of(resourcekey, blockposition, f, f1), true), false);
+            entityplayer.setRespawnPosition(new EntityPlayer.RespawnConfig(WorldData.a.of(resourcekey, blockposition, f, f1), true), false, org.bukkit.event.player.PlayerSpawnChangeEvent.Cause.COMMAND); // CraftBukkit
         }
 
         String s = resourcekey.identifier().toString();
