--- a/net/minecraft/server/commands/WeatherCommand.java
+++ b/net/minecraft/server/commands/WeatherCommand.java
@@ -33,11 +33,11 @@
     }
 
     private static int getDuration(CommandSourceStack source, int input, IntProvider defaultDistribution) {
-        return input == -1 ? defaultDistribution.sample(source.getServer().overworld().getRandom()) : input;
+        return input == -1 ? defaultDistribution.sample(source.getLevel().getRandom()) : input; // CraftBukkit - SPIGOT-7680: per-world
     }
 
     private static int setClear(CommandSourceStack source, int duration) {
-        source.getServer().overworld().setWeatherParameters(getDuration(source, duration, ServerLevel.RAIN_DELAY), 0, false, false);
+        source.getLevel().setWeatherParameters(getDuration(source, duration, ServerLevel.RAIN_DELAY), 0, false, false); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.clear");
         }, true);
@@ -45,7 +45,7 @@
     }
 
     private static int setRain(CommandSourceStack source, int duration) {
-        source.getServer().overworld().setWeatherParameters(0, getDuration(source, duration, ServerLevel.RAIN_DURATION), true, false);
+        source.getLevel().setWeatherParameters(0, getDuration(source, duration, ServerLevel.RAIN_DURATION), true, false); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.rain");
         }, true);
@@ -53,7 +53,7 @@
     }
 
     private static int setThunder(CommandSourceStack source, int duration) {
-        source.getServer().overworld().setWeatherParameters(0, getDuration(source, duration, ServerLevel.THUNDER_DURATION), true, true);
+        source.getLevel().setWeatherParameters(0, getDuration(source, duration, ServerLevel.THUNDER_DURATION), true, true); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.thunder");
         }, true);
