--- a/net/minecraft/world/entity/monster/skeleton/EntitySkeletonWither.java
+++ b/net/minecraft/world/entity/monster/skeleton/EntitySkeletonWither.java
@@ -94,7 +94,7 @@
             return false;
         } else {
             if (entity instanceof EntityLiving) {
-                ((EntityLiving) entity).addEffect(new MobEffect(MobEffects.WITHER, 200), this);
+                ((EntityLiving) entity).addEffect(new MobEffect(MobEffects.WITHER, 200), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
 
             return true;
