--- a/net/minecraft/world/entity/decoration/painting/Painting.java
+++ b/net/minecraft/world/entity/decoration/painting/Painting.java
@@ -60,7 +_,7 @@
         }
     }
 
-    private void setVariant(Holder<PaintingVariant> p_481528_) {
+    public void setVariant(Holder<PaintingVariant> p_481528_) {
         this.entityData.set(DATA_PAINTING_VARIANT_ID, p_481528_);
     }
 
@@ -163,7 +_,27 @@
         return AABB.ofSize(vec31, d2, d3, d4);
     }
 
-    private double offsetForPaintingSize(int p_480598_) {
+    public static AABB calculateBoundingBoxStatic(BlockPos blockpos, Direction direction, int width, int height) {
+        // CraftBukkit end
+        float f = 0.46875F;
+        Vec3 vec3 = Vec3.atCenterOf(blockpos).relative(direction, -0.46875D);
+        // CraftBukkit start
+        double d0 = offsetForPaintingSize(width);
+        double d1 = offsetForPaintingSize(height);
+        // CraftBukkit end
+        Direction direction1 = direction.getCounterClockWise();
+        Vec3 vec31 = vec3.relative(direction1, d0).relative(Direction.UP, d1);
+        Direction.Axis direction_axis = direction.getAxis();
+        // CraftBukkit start
+        double d2 = direction_axis == Direction.Axis.X ? 0.0625D : (double) width;
+        double d3 = (double) height;
+        double d4 = direction_axis == Direction.Axis.Z ? 0.0625D : (double) width;
+        // CraftBukkit end
+
+        return AABB.ofSize(vec31, d2, d3, d4);
+    }
+
+    private static double offsetForPaintingSize(int p_480598_) {
         return p_480598_ % 2 == 0 ? 0.5 : 0.0;
     }
 
