--- a/net/minecraft/world/item/ItemBoneMeal.java
+++ b/net/minecraft/world/item/ItemBoneMeal.java
@@ -35,6 +35,12 @@
 
     @Override
     public EnumInteractionResult useOn(ItemActionContext itemactioncontext) {
+        // CraftBukkit start - extract bonemeal application logic to separate, static method
+        return applyBonemeal(itemactioncontext);
+    }
+
+    public static EnumInteractionResult applyBonemeal(ItemActionContext itemactioncontext) {
+        // CraftBukkit end
         World world = itemactioncontext.getLevel();
         BlockPosition blockposition = itemactioncontext.getClickedPos();
         BlockPosition blockposition1 = blockposition.relative(itemactioncontext.getClickedFace());
@@ -42,7 +48,7 @@
 
         if (growCrop(itemstack, world, blockposition)) {
             if (!world.isClientSide()) {
-                itemstack.causeUseVibration(itemactioncontext.getPlayer(), GameEvent.ITEM_INTERACT_FINISH);
+                if (itemactioncontext.getPlayer() != null) itemstack.causeUseVibration(itemactioncontext.getPlayer(), GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
                 world.levelEvent(1505, blockposition, 15);
             }
 
@@ -53,7 +59,7 @@
 
             if (flag && growWaterPlant(itemstack, world, blockposition1, itemactioncontext.getClickedFace())) {
                 if (!world.isClientSide()) {
-                    itemstack.causeUseVibration(itemactioncontext.getPlayer(), GameEvent.ITEM_INTERACT_FINISH);
+                    if (itemactioncontext.getPlayer() != null) itemstack.causeUseVibration(itemactioncontext.getPlayer(), GameEvent.ITEM_INTERACT_FINISH); // CraftBukkit - SPIGOT-7518
                     world.levelEvent(1505, blockposition1, 15);
                 }
 
