--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -27,8 +_,9 @@
 import org.jspecify.annotations.Nullable;
 
 public abstract class BaseContainerBlockEntity extends BlockEntity implements Container, MenuProvider, Nameable {
-    private LockCode lockKey = LockCode.NO_LOCK;
-    private @Nullable Component name;
+    public LockCode lockKey = LockCode.NO_LOCK;
+    @Nullable
+    public Component name;
 
     protected BaseContainerBlockEntity(BlockEntityType<?> p_155076_, BlockPos p_155077_, BlockState p_155078_) {
         super(p_155076_, p_155077_, p_155078_);
@@ -118,9 +_,17 @@
 
     @Override
     public void setItem(int p_332705_, ItemStack p_332643_) {
+        setItem(p_332705_, p_332643_, false);
+    }
+
+    // Neo: Skip side-effects if insideTransaction is true so the caller can defer them until the transaction commits
+    @Override
+    public void setItem(int p_332705_, ItemStack p_332643_, boolean insideTransaction) {
         this.getItems().set(p_332705_, p_332643_);
         p_332643_.limitSize(this.getMaxStackSize(p_332643_));
-        this.setChanged();
+        if (!insideTransaction) {
+            this.setChanged();
+        }
     }
 
     @Override
