--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -26,7 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public abstract class BaseContainerBlockEntity extends BlockEntity implements Container, MenuProvider, Nameable {
-    private LockCode lockKey = LockCode.NO_LOCK;
+    public LockCode lockKey = LockCode.NO_LOCK;
     @Nullable
     private Component name;
 
