--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -82,13 +_,13 @@
     private static final boolean DEFAULT_MARKER = false;
     private boolean invisible = false;
     public long lastHit;
-    private int disabledSlots = 0;
-    private Rotations headPose = DEFAULT_HEAD_POSE;
-    private Rotations bodyPose = DEFAULT_BODY_POSE;
-    private Rotations leftArmPose = DEFAULT_LEFT_ARM_POSE;
-    private Rotations rightArmPose = DEFAULT_RIGHT_ARM_POSE;
-    private Rotations leftLegPose = DEFAULT_LEFT_LEG_POSE;
-    private Rotations rightLegPose = DEFAULT_RIGHT_LEG_POSE;
+    public int disabledSlots = 0;
+    public Rotations headPose = DEFAULT_HEAD_POSE;
+    public Rotations bodyPose = DEFAULT_BODY_POSE;
+    public Rotations leftArmPose = DEFAULT_LEFT_ARM_POSE;
+    public Rotations rightArmPose = DEFAULT_RIGHT_ARM_POSE;
+    public Rotations leftLegPose = DEFAULT_LEFT_LEG_POSE;
+    public Rotations rightLegPose = DEFAULT_RIGHT_LEG_POSE;
 
     public ArmorStand(EntityType<? extends ArmorStand> p_31553_, Level p_31554_) {
         super(p_31553_, p_31554_);
@@ -528,7 +_,7 @@
         return this.isMarker();
     }
 
-    private void setSmall(boolean p_31604_) {
+    public void setSmall(boolean p_31604_) {
         this.entityData.set(DATA_CLIENT_FLAGS, this.setBit(this.entityData.get(DATA_CLIENT_FLAGS), 1, p_31604_));
     }
 
@@ -552,7 +_,7 @@
         return (this.entityData.get(DATA_CLIENT_FLAGS) & 8) == 0;
     }
 
-    private void setMarker(boolean p_31682_) {
+    public void setMarker(boolean p_31682_) {
         this.entityData.set(DATA_CLIENT_FLAGS, this.setBit(this.entityData.get(DATA_CLIENT_FLAGS), 16, p_31682_));
     }
 
