--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -57,7 +_,7 @@
         this.entityData.set(DATA_DIRECTION, p_426128_);
     }
 
-    protected void setDirection(Direction p_31728_) {
+    public void setDirection(Direction p_31728_) {
         Objects.requireNonNull(p_31728_);
         Validate.isTrue(p_31728_.getAxis().isHorizontal());
         this.setDirectionRaw(p_31728_);
@@ -83,7 +_,7 @@
         if (this.hasLevelCollision(this.getPopBox())) {
             return false;
         } else {
-            boolean flag = BlockPos.betweenClosedStream(this.calculateSupportBox()).allMatch(p_423325_ -> {
+            boolean flag = BlockPos.betweenClosedStream(this.calculateSupportBox()).filter(pos -> !net.minecraft.world.level.block.Block.canSupportCenter(this.level(), pos, this.getDirection())).allMatch(p_423325_ -> {
                 BlockState blockstate = this.level().getBlockState(p_423325_);
                 return blockstate.isSolid() || DiodeBlock.isDiode(blockstate);
             });
