--- a/net/minecraft/network/syncher/DataWatcher.java
+++ b/net/minecraft/network/syncher/DataWatcher.java
@@ -50,7 +50,7 @@
     }
 
     private <T> DataWatcher.Item<T> getItem(DataWatcherObject<T> datawatcherobject) {
-        return this.itemsById[datawatcherobject.id()];
+        return (DataWatcher.Item<T>) this.itemsById[datawatcherobject.id()]; // CraftBukkit - decompile error
     }
 
     public <T> T get(DataWatcherObject<T> datawatcherobject) {
@@ -73,6 +73,13 @@
 
     }
 
+    // CraftBukkit start - add method from above
+    public <T> void markDirty(DataWatcherObject<T> datawatcherobject) {
+        this.getItem(datawatcherobject).setDirty(true);
+        this.isDirty = true;
+    }
+    // CraftBukkit end
+
     public boolean isDirty() {
         return this.isDirty;
     }
@@ -126,7 +133,7 @@
         if (!Objects.equals(datawatcher_c.serializer(), datawatcher_item.accessor.serializer())) {
             throw new IllegalStateException(String.format(Locale.ROOT, "Invalid entity data item type for field %d on entity %s: old=%s(%s), new=%s(%s)", datawatcher_item.accessor.id(), this.entity, datawatcher_item.value, datawatcher_item.value.getClass(), datawatcher_c.value, datawatcher_c.value.getClass()));
         } else {
-            datawatcher_item.setValue(datawatcher_c.value);
+            datawatcher_item.setValue((T) datawatcher_c.value); // CraftBukkit - decompile error
         }
     }
 
@@ -211,7 +218,7 @@
     public static class a {
 
         private final SyncedDataHolder entity;
-        private final @Nullable DataWatcher.Item<?>[] itemsById;
+        private final DataWatcher.Item<?> @Nullable [] itemsById; // CraftBukkit - decompile error
 
         public a(SyncedDataHolder synceddataholder) {
             this.entity = synceddataholder;
