--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
@@ -59,12 +_,12 @@
     private static final int MAX_MOB_TRACKING_DISTANCE_SQR = Mth.square(47);
     private static final float SPAWNING_AMBIENT_SOUND_CHANCE = 0.02F;
     private final TrialSpawnerStateData data = new TrialSpawnerStateData();
-    private TrialSpawner.FullConfig config;
+    public TrialSpawner.FullConfig config;
     private final TrialSpawner.StateAccessor stateAccessor;
     private PlayerDetector playerDetector;
     private final PlayerDetector.EntitySelector entitySelector;
     private boolean overridePeacefulAndMobSpawnRule;
-    private boolean isOminous;
+    public boolean isOminous;
 
     public TrialSpawner(
         TrialSpawner.FullConfig p_422371_, TrialSpawner.StateAccessor p_312198_, PlayerDetector p_312690_, PlayerDetector.EntitySelector p_323834_
@@ -212,9 +_,9 @@
                 }
 
                 boolean flag = spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().getString("id").isPresent();
-                if (flag) {
-                    mob.finalizeSpawn(p_312582_, p_312582_.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.TRIAL_SPAWNER, null);
-                }
+                // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                // The local flag is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, p_312582_, p_312582_.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.TRIAL_SPAWNER, null, this, flag);
 
                 mob.setPersistenceRequired();
                 spawndata.getEquipment().ifPresent(mob::equip);
@@ -414,5 +_,13 @@
         TrialSpawnerState getState();
 
         void markUpdated();
+    }
+
+    @Override
+    public com.mojang.datafixers.util.@org.jspecify.annotations.Nullable Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        if (this.stateAccessor instanceof net.minecraft.world.level.block.entity.TrialSpawnerBlockEntity be) {
+            return com.mojang.datafixers.util.Either.left(be);
+        }
+        return null;
     }
 }
